crates/red_knot_test/src/assertion.rs

```diff
@@ -227,50 +226,78 @@ pub(crate) struct LineAssertions<'a> {
 }
 
 impl<'a> Deref for LineAssertions<'a> {
-    type Target = [Assertion<'a>];
+    type Target = [UnparsedAssertion<'a>];
 
     fn deref(&self) -> &Self::Target {
         &self.assertions
     }
 }
 
-static TYPE_RE: LazyLock<Regex> =
-    LazyLock::new(|| Regex::new(r"^#\s*revealed:\s*(?<ty_display>.+?)\s*$").unwrap());
-
-static ERROR_RE: LazyLock<Regex> = LazyLock::new(|| {
-    Regex::new(
-        r#"^#\s*error:(\s*(?<column>\d+))?(\s*\[(?<rule>.+?)\])?(\s*"(?<message>.+?)")?\s*$"#,
-    )
-    .unwrap()
-});
+const TYPE_PREFIX: &str = "# revealed:";
+const ERROR_PREFIX: &str = "# error:";
 
 /// A single diagnostic assertion comment.
+///
+/// This type represents an *attempted* assertion, but not necessarily a *valid* assertion.
+/// Parsing is done lazily in `matcher.rs`; this allows us to emit nicer error messages
+/// in the event of an invalid assertion
 #[derive(Debug)]
-pub(crate) enum Assertion<'a> {
-    /// A `revealed: ` assertion.
+pub(crate) enum UnparsedAssertion<'a> {
+    /// A `# revealed:` assertion.
     Revealed(&'a str),
 
-    /// An `error: ` assertion.
-    Error(ErrorAssertion<'a>),
+    /// An `# error:` assertion.
+    Error(&'a str),
 }
 
-impl<'a> Assertion<'a> {
+impl<'a> UnparsedAssertion<'a> {
+    /// Returns `Some(_)` if the comment starts with `# error:` or `# revealed:`,
+    /// indicating that it is an assertion comment.
     fn from_comment(comment: &'a str) -> Option<Self> {
-        if let Some(caps) = TYPE_RE.captures(comment) {
-            Some(Self::Revealed(caps.name("ty_display").unwrap().as_str()))
-        } else {
-            ERROR_RE.captures(comment).map(|caps| {
-                Self::Error(ErrorAssertion {
-                    rule: caps.name("rule").map(|m| m.as_str()),
-                    column: caps.name("column").and_then(|m| m.as_str().parse().ok()),
-                    message_contains: caps.name("message").map(|m| m.as_str()),
-                })
-            })
+        comment
+            .strip_prefix(TYPE_PREFIX)
+            .map(Self::Revealed)
+            .or_else(|| comment.strip_prefix(ERROR_PREFIX).map(Self::Error))
```

Comment:
Not necessarily something for this PR but two common errors that I made in the past:

* I wrote `error[code]:` which is wrong
* I misspelled `error` or `revealed`
* Too much or not enough whitespace between `#` and the pragma keyword` 

The second one is probably easy to detect (error on any comment of the form `<pragma>:` and allowlist legit pragmas (`fmt`, `type`, `knot`)

Reply:
> The second one is probably easy to detect (error on any comment of the form `<pragma>:` and allowlist legit pragmas (`fmt`, `type`, `knot`)

Hah, you would _think_ so... I've just spent a bit of time playing around with it, and it seems surprisingly hard to shoehorn into our current architecture :-(

I'm making the whitespace handling more lenientm though; that's easy to fix